util/stats/query_stats.go

```diff
@@ -145,14 +219,111 @@ func (s *SpanTimer) Finish() {
 	}
 }
 
+type Statistics struct {
+	Timers  *QueryTimers
+	Samples *QuerySamples
+}
+
 type QueryTimers struct {
 	*TimerGroup
 }
 
+type TotalSamplesPerTime map[int64]int
+
+type QuerySamples struct {
+	// PeakSamples represent the highest count of samples considered
+	// while evaluating a query. It corresponds to the peak value of
+	// currentSamples, which is in turn compared against the MaxSamples
+	// configured in the engine.
+	PeakSamples int
+
+	// TotalSamplesPerTime represents the total number of samples scanned
+	// while evaluating a query.
+	TotalSamples        int
+	TotalSamplesPerTime []int
+
+	EnablePerStepStats bool
+	startTimestamp     int64
+	interval           int64
+}
+
+type Stats struct {
+	TimerStats  *QueryTimers
+	SampleStats *QuerySamples
```

Comment:
If we are concerned about impact, we could later on make that an interface, so when it is disabled we have a NoopSampleStats that has lesser impact. I see @beorn7 has launched prombench already, it might not be needed.